cmd/object-api-utils.go

```diff
@@ -302,3 +303,36 @@ type byBucketName []BucketInfo
 func (d byBucketName) Len() int           { return len(d) }
 func (d byBucketName) Swap(i, j int)      { d[i], d[j] = d[j], d[i] }
 func (d byBucketName) Less(i, j int) bool { return d[i].Name < d[j].Name }
+
+// GetObjectReader is a type that wraps a reader with a lock to
+// provide a ReadCloser interface that unlocks on Close()
+type GetObjectReader struct {
+	lock RWLocker
+	pr   io.Reader
+
+	// register any clean up actions (happens before unlocking)
+	cleanUp func()
+}
+
+// NewGetObjectReader creates a new GetObjectReader
+func NewGetObjectReader(reader io.Reader, lock RWLocker, cleanUp func()) GetObjectReader {
+	return GetObjectReader{
+		lock:    lock,
+		pr:      reader,
+		cleanUp: cleanUp,
+	}
+}
+
+// Close - calls unlock and any registered clean up action
+func (g *GetObjectReader) Close() error {
+	if g.lock != nil {
+		g.lock.RUnlock()
```

Comment:
Correct me if I'm wrong but shouldn't `g.cleanUp` be called here first? And if we call a cleanup (which may panic / `logger.CriticalIf` we may should use `defer g.lock.RUnlock()` to be sure that the lock is released.

Reply:
Yes, this was the intention, I guess i missed it while refactoring my own PR.